File: 이훈일/Programmers/Level2/쿼드압축 후 개수 세기.java
class Solution {

    int[] result = new int[]{0,0};

    public int[] solution(int[][] arr) {
        int len = arr.length;
        division(arr,0 ,0 ,len);
        return this.result;
    }

    public void division(int[][] arr, int x, int y, int q) {
        int start = arr[x][y];
        int bool = 2;

        if (q == 1){
            this.result[start] += 1;
            return;
        }

        for(int i = x; i < x+q; i++) {
            for(int j = y; j < y+q; j++) {
                if(start != arr[i][j]) {
                    bool += 1;
                    break;
                }
            }
        }

        if(bool == 2) {
            this.result[start] += 1;
        } else {
            q = (Integer)q/2;
            division(arr, x, y, q);
            division(arr, x, y+q, q);
            division(arr, x+q, y, q);
            division(arr, x+q, y+q, q);
        }
    }
}

File: 이건희/level2/쿼드압축 후 개수 세기.py
def solution(arr):
    result = [0, 0]
    num = len(arr)

    def quad(arr, num):
        nonlocal result

        now = arr[0][0]

        if num == 1:
            result[now] += 1
            return
        
        for i in arr:
            if i.count(now) != num:
                break
        else:
            result[now] += 1
            return
        
        num //= 2
        
        quad([i[:num] for i in arr[:num]], num)
        quad([i[num:] for i in arr[:num]], num)
        quad([i[:num] for i in arr[num:]], num)
        quad([i[num:] for i in arr[num:]], num)
        
    quad(arr, num)
    return result
